server/src/main/java/org/elasticsearch/common/util/CancellableSingleObjectCache.java

```diff
@@ -140,25 +152,17 @@ public final void get(Input input, BooleanSupplier isCancelled, ActionListener<V
                 if (currentCachedItem != null) {
                     currentCachedItem.decRef();
                 }
-                startRefresh(input, newCachedItem);
                 final boolean listenerAdded = newCachedItem.addListener(listener, isCancelled);
                 assert listenerAdded;
-                newCachedItem.decRef();
+                newCachedItem.decRef(); // release our ref before calling refresh so that we're not blocking a cancellation
+                newCachedItem.startRefresh(input);
```

Comment:
This bit is really a bugfix and deserves backport to 7.16. I think I might want to rely on the supersede functionality in 7.16 for fixing #79104 too. We can wait until then to backport if you prefer but I would also think this could be ok to backport now.

Reply:
:+1: I also think this is good to backport, I'll do so.